# Introducción a Pandas

## Objetivo:

Explorar las funciones básicas de la biblioteca Pandas en Python para cargar, visualizar y manipular estructuras de datos tipo DataFrame.

*Nota: Este laboratorio está diseñado para desarrollar razonamiento estadístico. Evita el uso de herramientas de IA para responder las preguntas a menos que esto sea explícitamente sugerido* 🤖. *En su lugar, enfócate en aplicar los conceptos revisados.*

## Actividad 1

Antes de comenzar a trabajar con bases de datos más complejas, aprenderás a crear tus propios datos dentro de Python usando Pandas. Esto te permitirá entender mejor cómo funciona un DataFrame y practicar operaciones básicas de conteo y selección.

1. Crea un DataFrame con los siguientes datos de estudiantes:

<center>

|Nombre| Edad (años)| Carrera|
|------|:-----------:|--------|
|Rodrigo | 21 | Informática|
|Pedro | 23 | Telecomunicaciones|
|Ignacia | 20 | Informática|
|Luis | 24 | Informática|
|Valentina | 22 | Analista|
|Carlos | 20 | Telecomunicaciones|
|Martina | 22 | Analista|
|Javiera | 21 | Telecomunicaciones|
|Camilo | 20 | Informática|
|Bastián | 20 | Informática|

</center>

Luego, muestra en la consola el contenido completo del DataFrame.

In [22]:
#Primero importamos la biblioteca Pandas que nos permitirá trabajar con DataFrames
import pandas as pd

In [1]:
#Creamos un DataFrame con la información de la tabla
tabla_1 = pd.DataFrame({
    'Nombre': ['Rodrigo', 'Pedro', 'Ignacia', 'Luis', 'Valentina', 'Carlos',
               'Martina', 'Javiera', 'Camilo', 'Bastián'],
    'Edad (años)': [21, 23, 20, 24, 22, 20, 22, 21, 20, 20],
    'Carrera': ['Informática', 'Telecomunicaciones', 'Informática', 'Informática', 'Analista',
                'Telecomunicaciones', 'Analista', 'Telecomunicaciones', 'Informática', 'Informática']
})

#Mostramos el DataFrame completo
tabla_1

Nombre  Edad (años)             Carrera
0    Rodrigo           21         Informática
1      Pedro           23  Telecomunicaciones
2    Ignacia           20         Informática
3       Luis           24         Informática
4  Valentina           22            Analista
5     Carlos           20  Telecomunicaciones
6    Martina           22            Analista
7    Javiera           21  Telecomunicaciones
8     Camilo           20         Informática
9    Bastián           20         Informática

2. Muestra únicamente la columna "Carrera" del DataFrame creado en el inciso anterior.

In [2]:
#Mostramos solo la columna "Carrera" del DataFrame
tabla_1['Carrera']

0           Informática
1    Telecomunicaciones
2           Informática
3           Informática
4              Analista
5    Telecomunicaciones
6              Analista
7    Telecomunicaciones
8           Informática
9           Informática
Name: Carrera, dtype: object

3. ¿Cuántos estudiantes hay en cada carrera? Utiliza alguna función de Python para responder.

In [3]:
#Contamos cuántas veces aparece cada carrera en la columna "Carrera"
cant_estudiantes = tabla_1['Carrera'].value_counts()

#Mostramos el resultado de aplicar la función value_counts()
cant_estudiantes

Carrera
Informática           5
Telecomunicaciones    3
Analista              2
Name: count, dtype: int64

De los estudiantes indicados en la tabla inicial, 5 pertenecen a la carrera de Informática, 3 a Telecomunicaciones y solo 2 a Analista.

4. ¿Cuál es la edad más común entre los estudiantes?

In [4]:
#Contamos cuántas veces aparece cada edad en la columna "Edad (años)"
edad_comun = tabla_1['Edad (años)'].value_counts()

#Mostramos el resultado de aplicar la función value_counts()
edad_comun

Edad (años)
20    4
21    2
22    2
23    1
24    1
Name: count, dtype: int64

La edad más común de este grupo de estudiantes es 20 años, con 4 personas.

5. Se espera que estos estudiantes egresen de su carrera en aproximadamente 3 años más. Agrega una nueva columna al DataFrame llamada "Edad al egresar" que contenga la edad que tendrá cada estudiante al finalizar sus estudios. Luego, muestra en la consola el DataFrame actualizado.

In [5]:
#Agregamos una nueva columna al DataFrame inicial con la suma indicada
tabla_1['Edad al egresar'] =  tabla_1['Edad (años)'] + 3

#Mostramos el DataFrame actualizado
tabla_1

Nombre  Edad (años)             Carrera  Edad al egresar
0    Rodrigo           21         Informática               24
1      Pedro           23  Telecomunicaciones               26
2    Ignacia           20         Informática               23
3       Luis           24         Informática               27
4  Valentina           22            Analista               25
5     Carlos           20  Telecomunicaciones               23
6    Martina           22            Analista               25
7    Javiera           21  Telecomunicaciones               24
8     Camilo           20         Informática               23
9    Bastián           20         Informática               23

## Actividad 2

Eres parte del equipo de soporte de datos de una empresa de telecomunicaciones. Te han pedido cargar una base de datos de clientes y realizar una primera revisión para comprender su estructura y conocer su contenido.

Descarga la base de datos **<a href="https://docs.google.com/spreadsheets/d/1v6euDfqECe1HTJz0yFUrn7mmrt-rG9gG/export?format=xlsx
">aquí</a>** y sigue las instrucciones detalladas a continuación:

1. Sube la base de datos al entorno "Archivos" de Google Colab (ícono de carpeta en la barra lateral izquierda). Luego, cárgala en un DataFrame llamado `df_1`.

$\hspace{1 cm}$ *Nota: Recuerda que para cargar la base de datos a un DF debes utilizar la biblioteca Pandas de Python*.

In [6]:
import pandas as pd
df_1 =  pd.read_excel('Base0.xlsx')

2. Explora el contenido del DataFrame mostrando las primeras y últimas 10 filas.

In [7]:
#Primeras 10 filas
df_1.head(10)

ID     Género  Ciudadano mayor Pareja Dependientes  Antigüedad  \
0  7590-VHVEG   Femenino                0     Sí           No           1   
1  5575-GNVDE  Masculino                0     No           No          34   
2  3668-QPYBK  Masculino                0     No           No           2   
3  7795-CFOCW  Masculino                0     No           No          45   
4  9237-HQITU   Femenino                0     No           No           2   
5  9305-CDSKC   Femenino                0     No           No           8   
6  1452-KIOVK  Masculino                0     No           Sí          22   
7  6713-OKOMC   Femenino                0     No           No          10   
8  7892-POOKP   Femenino                0     Sí           No          28   
9  6388-TABGU  Masculino                0     No           Sí          62   

  Servicio telefónico            Líneas múltiples Servicio de Internet  \
0                  No  No hay servicio telefónico                  DSL   
1                  Sí                          No                  DSL   
2                  Sí                          No                  DSL   
3                  No  No hay servicio telefónico                  DSL   
4                  Sí                          No         Fibra óptica   
5                  Sí                          Sí         Fibra óptica   
6                  Sí                          Sí         Fibra óptica   
7                  No  No hay servicio telefónico                  DSL   
8                  Sí                          Sí         Fibra óptica   
9                  Sí                          No                  DSL   

  Protección contra Malware Protección del dispositivo Soporte técnico  \
0                        No                         No              No   
1                        Sí                         Sí              No   
2                        Sí                         No              No   
3                        Sí                         Sí              Sí   
4                        No                         No              No   
5                        No                         Sí              No   
6                        No                         No              No   
7                        Sí                         No              No   
8                        No                         Sí              Sí   
9                        Sí                         No              No   

  Televisión en streaming Películas en streaming   Contrato  \
0                      No                     No  Mes a mes   
1                      No                     No     Un año   
2                      No                     No  Mes a mes   
3                      No                     No     Un año   
4                      No                     No  Mes a mes   
5                      Sí                     Sí  Mes a mes   
6                      Sí                     No  Mes a mes   
7                      No                     No  Mes a mes   
8                      Sí                     Sí  Mes a mes   
9                      No                     No     Un año   

  Facturación sin papel                       Método de pago  \
0                    Sí                   Cheque electrónico   
1                    No            Cheque enviado por correo   
2                    Sí            Cheque enviado por correo   
3                    No  Transferencia bancaria (automática)   
4                    Sí                   Cheque electrónico   
5                    Sí                   Cheque electrónico   
6                    Sí      Tarjeta de crédito (automática)   
7                    No            Cheque enviado por correo   
8                    Sí                   Cheque electrónico   
9                    No  Transferencia bancaria (automática)   

   Cargos mensuales (dólares)  Cargos totales  (dólares) Deserción Cliente  
0                       29.85                      29.85 

In [8]:
#Últimas 10 filas
df_1.tail(10)

ID     Género  Ciudadano mayor Pareja Dependientes  Antigüedad  \
4089  0829-XXPLX   Femenino                0     No           No          20   
4090  9974-JFBHQ  Masculino                0     No           Sí          64   
4091  5356-RHIPP  Masculino                0     No           No           1   
4092  6624-JDRDS   Femenino                0     No           No           6   
4093  0608-JDVEC  Masculino                0     Sí           No          50   
4094  9780-FKVVF  Masculino                0     No           No           6   
4095  1919-RTPQD  Masculino                0     Sí           Sí           7   
4096  5214-NLTIT  Masculino                0     Sí           Sí          72   
4097  3345-PBBFH  Masculino                0     Sí           No           8   
4098  5055-BRMNE   Femenino                0     Sí           Sí          67   

     Servicio telefónico            Líneas múltiples Servicio de Internet  \
4089                  Sí                          No         Fibra óptica   
4090                  Sí                          No         Fibra óptica   
4091                  Sí                          No                   No   
4092                  No  No hay servicio telefónico                  DSL   
4093                  Sí                          No                   No   
4094                  Sí                          No                  DSL   
4095                  Sí                          No                  DSL   
4096                  Sí                          Sí         Fibra óptica   
4097                  Sí                          No                  DSL   
4098                  Sí                          Sí         Fibra óptica   

     Protección contra Malware Protección del dispositivo  \
4089                        Sí                         Sí   
4090                        No                         No   
4091  Sin servicio de internet   Sin servicio de internet   
4092                        Sí                         No   
4093  Sin servicio de internet   Sin servicio de internet   
4094                        Sí                         No   
4095                        No                         No   
4096                        No                         Sí   
4097                        No                         No   
4098                        Sí                         No   

               Soporte técnico   Televisión en streaming  \
4089                        No                        Sí   
4090                        Sí                        Sí   
4091  Sin servicio de internet  Sin servicio de internet   
4092                        No                        No   
4093  Sin servicio de internet  Sin servicio de internet   
4094                        No                        No   
4095                        No                        No   
4096                        No                        No   
4097                        No                        No   
4098                        Sí                        Sí   

        Películas en streaming   Contrato Facturación sin papel  \
4089                        No  Mes a mes                    Sí   
4090                        Sí  Mes a mes                    Sí   
4091  Sin servicio de internet  Mes a mes                    Sí   
4092                        No  Mes a mes                    No   
4093  Sin servicio de internet   Dos años                    No   
4094                        Sí  Mes a mes                    Sí   
4095                        No  Mes a mes                    No   
4096                        Sí   Dos años                    Sí   
4097                        No  Mes a mes                    Sí   
4098                        Sí   Dos años                    No   

                           Método de pago  Cargos mensuales (dólares)  \
4089  Transferencia bancaria (automática)                       89.40   
4090      Tarjeta de crédito (automática)                       99.1

3. ¿Cuántas filas y columnas tiene la base de datos? Responde utilizando solo una línea de código.

In [9]:
#Para conocer la dimensión del DataFrame usaremos la función shape
df_1.shape

(4099, 20)

La base de datos tiene 4099 filas y 20 columnas.

4. ¿Qué tipo de objeto es `df_1`? Justifica utilizando una función de Python.

In [10]:
type(df_1)

pandas.core.frame.DataFrame

Tal como indica la función `type()`, el objeto `df_1` es un DataFrame, esto quiere decir que es una tabla de datos organizada en filas y columnas.

*Nota: Google Colab puede mostrar información adicional del objeto analizado con `type()`. En este caso, además del tipo de objeto,  muestra su definición y descripción técnica.*

5. Usa la función `info()` para examinar el contenido  del DataFrame. ¿Qué tipo de datos contiene cada columna?

In [11]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4099 entries, 0 to 4098
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          4099 non-null   object 
 1   Género                      4099 non-null   object 
 2   Ciudadano mayor             4099 non-null   int64  
 3   Pareja                      4099 non-null   object 
 4   Dependientes                4099 non-null   object 
 5   Antigüedad                  4099 non-null   int64  
 6   Servicio telefónico         4099 non-null   object 
 7   Líneas múltiples            4099 non-null   object 
 8   Servicio de Internet        4099 non-null   object 
 9   Protección contra Malware   4099 non-null   object 
 10  Protección del dispositivo  4099 non-null   object 
 11  Soporte técnico             4099 non-null   object 
 12  Televisión en streaming     4099 non-null   object 
 13  Películas en streaming      4099 

El DataFrame contiene 2 columnas con datos tipo `float` (números decimales), 2 columnas con datos tipo `int` (números enteros) y 16 columnas con datos tipo  `object`, que generalmente corresponden a datos que almacenan texto.

6. Cambia el tipo de dato de la columna "Antigüedad" a `float`. Para lograrlo, deberás reasignar esta columna con el nuevo tipo de dato. Luego, muestra las primeras 3 filas del DataFrame.

In [12]:
#Accedemos a la columna "Antigüedad" y la reasignamos con el cambio utilizando la
#función astype()
df_1['Antigüedad'] = df_1['Antigüedad'].astype(float)

#Mostramos las primeras 3 filas
df_1.head(3)

ID     Género  Ciudadano mayor Pareja Dependientes  Antigüedad  \
0  7590-VHVEG   Femenino                0     Sí           No         1.0   
1  5575-GNVDE  Masculino                0     No           No        34.0   
2  3668-QPYBK  Masculino                0     No           No         2.0   

  Servicio telefónico            Líneas múltiples Servicio de Internet  \
0                  No  No hay servicio telefónico                  DSL   
1                  Sí                          No                  DSL   
2                  Sí                          No                  DSL   

  Protección contra Malware Protección del dispositivo Soporte técnico  \
0                        No                         No              No   
1                        Sí                         Sí              No   
2                        Sí                         No              No   

  Televisión en streaming Películas en streaming   Contrato  \
0                      No                     No  Mes a mes   
1                      No                     No     Un año   
2                      No                     No  Mes a mes   

  Facturación sin papel             Método de pago  \
0                    Sí         Cheque electrónico   
1                    No  Cheque enviado por correo   
2                    Sí  Cheque enviado por correo   

   Cargos mensuales (dólares)  Cargos totales  (dólares) Deserción Cliente  
0                       29.85                      29.85                No  
1                       56.95                    1889.50                No  
2                       53.85                     108.15                Sí

7. Verifica que la columna "Antigüedad" haya cambiado correctamente su tipo de dato a `float`.

In [13]:
#Verificamos el cambio realizado con la función info()
df_1['Antigüedad'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 4099 entries, 0 to 4098
Series name: Antigüedad
Non-Null Count  Dtype  
--------------  -----  
4099 non-null   float64
dtypes: float64(1)
memory usage: 32.2 KB


8. ¿Te parece correcto cambiar el tipo de dato de la columna "Cargos mensuales (dólares)" a tipo `int`? Justifica tu respuesta.

No me parece correcto cambiar el tipo de dato de la columna "Cargos mensuales (dólares)", ya que esta columna representa montos de dinero expresados con valores decimales. Al realizar el cambio, se perdería precisión, lo que afectaría los análisis posteriores.

## Actividad 3

La empresa de telecomunicaciones te ha pedido que analices ciertas variables clave de sus clientes. Necesitas extraer columnas específicas, aplicar filtros básicos y comenzar a preparar los datos para futuros análisis más avanzados.

Puedes descargar la base de datos **<a href="https://docs.google.com/spreadsheets/d/1v6euDfqECe1HTJz0yFUrn7mmrt-rG9gG/export?format=xlsx
">aquí</a>** para responder las siguientes preguntas:

1. Muestra las cuatro primeras filas de las columnas "Género", "Antigüedad" y "Deserción Cliente" del DataFrame.

In [14]:
import pandas as pd
df_1 = pd.read_excel('Base0.xlsx')

In [15]:
#Vamos a mostrar solo las columnas solicitadas
df_1[['Género', 'Antigüedad', 'Deserción Cliente']].head(4)

Género  Antigüedad Deserción Cliente
0   Femenino           1                No
1  Masculino          34                No
2  Masculino           2                Sí
3  Masculino          45                No

2. ¿Cuántos clientes de la base de datos pertenecen al género masculino y cuántos al femenino?

In [16]:
#Vamos a contar la cantidad de hombres y mujeres de la base de datos.
#Para esto, agrupamos con la función groupby por los valores de la columna "Género"
#y luego usamos esa misma columna para contar cuántos registros hay en cada grupo.
genero =  df_1.groupby('Género')['Género'].count()

#Mostramos la variable "genero"
genero

Género
Femenino     2016
Masculino    2083
Name: Género, dtype: int64

En la base de datos se registran 2.016 clientes de género femenino y 2.083 de género masculino.

3. Calcula el total que un cliente habría pagado si su cargo mensual se hubiera mantenido constante durante todo el tiempo en la empresa. Para esto, agrega una nueva columna al DataFrame llamada "Cargos acumulados (dólares)", que sea el resultado de multiplicar "Cargos mensuales" por "Antigüedad". Luego, muestra las primeras 5 filas del DataFrame para verificar que la nueva columna fue agregada correctamente.

In [17]:
#Agregamos la nueva columna
df_1['Cargos acumulados (dólares)'] = df_1['Cargos mensuales (dólares)'] * df_1['Antigüedad']

#Mostramos las primeras 5 filas de DataFrame
df_1.head()

ID     Género  Ciudadano mayor Pareja Dependientes  Antigüedad  \
0  7590-VHVEG   Femenino                0     Sí           No           1   
1  5575-GNVDE  Masculino                0     No           No          34   
2  3668-QPYBK  Masculino                0     No           No           2   
3  7795-CFOCW  Masculino                0     No           No          45   
4  9237-HQITU   Femenino                0     No           No           2   

  Servicio telefónico            Líneas múltiples Servicio de Internet  \
0                  No  No hay servicio telefónico                  DSL   
1                  Sí                          No                  DSL   
2                  Sí                          No                  DSL   
3                  No  No hay servicio telefónico                  DSL   
4                  Sí                          No         Fibra óptica   

  Protección contra Malware  ... Soporte técnico Televisión en streaming  \
0                        No  ...              No                      No   
1                        Sí  ...              No                      No   
2                        Sí  ...              No                      No   
3                        Sí  ...              Sí                      No   
4                        No  ...              No                      No   

  Películas en streaming   Contrato Facturación sin papel  \
0                     No  Mes a mes                    Sí   
1                     No     Un año                    No   
2                     No  Mes a mes                    Sí   
3                     No     Un año                    No   
4                     No  Mes a mes                    Sí   

                        Método de pago Cargos mensuales (dólares)  \
0                   Cheque electrónico                      29.85   
1            Cheque enviado por correo                      56.95   
2            Cheque enviado por correo                      53.85   
3  Transferencia bancaria (automática)                      42.30   
4                   Cheque electrónico                      70.70   

   Cargos totales  (dólares)  Deserción Cliente Cargos acumulados (dólares)  
0                      29.85                 No                       29.85  
1                    1889.50                 No                     1936.30  
2                     108.15                 Sí                      107.70  
3                    1840.75                 No                     1903.50  
4                     151.65                 Sí                      141.40  

[5 rows x 21 columns]

4. Crea un nuevo DataFrame que contenga la información solo de los clientes cuyo "Servicio de Internet" es Fibra óptica. ¿Cuántas filas tiene este nuevo DataFrame? Muestra sus primeras 5 filas para comprobar que el filtro se aplicó correctamente.

In [18]:
#Vamos a filtrar el DataFrame y guardar el contenido en una nueva variable llamada df_filtrado.
#Para esto, seleccionaremos todas las filas de df_1 cuyo valor de
#"Servicio de Internet" sea igual a "Fibra óptica"
df_filtrado = df_1[df_1['Servicio de Internet'] == 'Fibra óptica']

#Ahora calculamos la cantidad de filas de este nuevo DataFrame
cant_filas = df_filtrado.shape[0]

print(f'Cantidad de filas: {cant_filas}.')

#Mostramos las primeras 5 filas del DataFrame filtrado
df_filtrado.head()

Cantidad de filas: 1804.


ID     Género  Ciudadano mayor Pareja Dependientes  Antigüedad  \
4   9237-HQITU   Femenino                0     No           No           2   
5   9305-CDSKC   Femenino                0     No           No           8   
6   1452-KIOVK  Masculino                0     No           Sí          22   
8   7892-POOKP   Femenino                0     Sí           No          28   
12  8091-TTVAX  Masculino                0     Sí           No          58   

   Servicio telefónico Líneas múltiples Servicio de Internet  \
4                   Sí               No         Fibra óptica   
5                   Sí               Sí         Fibra óptica   
6                   Sí               Sí         Fibra óptica   
8                   Sí               Sí         Fibra óptica   
12                  Sí               Sí         Fibra óptica   

   Protección contra Malware  ... Soporte técnico Televisión en streaming  \
4                         No  ...              No                      No   
5                         No  ...              No                      Sí   
6                         No  ...              No                      Sí   
8                         No  ...              Sí                      Sí   
12                        No  ...              No                      Sí   

   Películas en streaming   Contrato Facturación sin papel  \
4                      No  Mes a mes                    Sí   
5                      Sí  Mes a mes                    Sí   
6                      No  Mes a mes                    Sí   
8                      Sí  Mes a mes                    Sí   
12                     Sí     Un año                    No   

                     Método de pago Cargos mensuales (dólares)  \
4                Cheque electrónico                      70.70   
5                Cheque electrónico                      99.65   
6   Tarjeta de crédito (automática)                      89.10   
8                Cheque electrónico                     104.80   
12  Tarjeta de crédito (automática)                     100.35   

    Cargos totales  (dólares)  Deserción Cliente Cargos acumulados (dólares)  
4                      151.65                 Sí                       141.4  
5                      820.50                 Sí                       797.2  
6                     1949.40                 No                      1960.2  
8                     3046.05                 Sí                      2934.4  
12                    5681.10                 No                      5820.3  

[5 rows x 21 columns]

El nuevo DataFrame contiene solo a los clientes cuyo servicio de internet es "Fibra óptica", por esto tiene menos filas que el DataFrame original, 1804 comparado con 4099<sup>1</sup>.

Al aplicar el filtro, los índices que se mantienen son los mismos que el DataFrame original `df_1`, por eso no son correlativos ni comienzan desde cero. Esto permite mantener trazabilidad con la fuente original de los datos.

Si deseas que el nuevo DataFrame tenga índices consecutivos desde cero, se puede usar la función `reset_index(drop = True)`.


<br>


<small><sup>1</sup> *Este valor se obtuvo en la Actividad 2, inciso 3.*</small>

5. ¿Cuántos clientes con fibra óptica abandonaron la empresa? Muestra las primeras 5 filas del nuevo DataFrame para comprobar que el filtro se aplicó correctamente.

In [19]:
#Para responder la pregunta, se debe trabajar con el DataFrame filtrado del inciso anterior y filtrarlo nuevamente.
#Esta vez, seleccionamos todas las filas donde el valor de Deserción Cliente sea Sí.
df_filtrado2 = df_filtrado[df_filtrado['Deserción Cliente'] == 'Sí']

#Calculamos la cantidad de filas de este nuevo DataFrame
desercion_si = df_filtrado2.shape[0]

print(f'Cantidad de filas: {desercion_si}')

#Mostramos las primeras cinco filas
df_filtrado2.head()

Cantidad de filas: 760


ID     Género  Ciudadano mayor Pareja Dependientes  Antigüedad  \
4   9237-HQITU   Femenino                0     No           No           2   
5   9305-CDSKC   Femenino                0     No           No           8   
8   7892-POOKP   Femenino                0     Sí           No          28   
13  0280-XJGEX  Masculino                0     No           No          49   
26  6467-CHFZW  Masculino                0     Sí           Sí          47   

   Servicio telefónico Líneas múltiples Servicio de Internet  \
4                   Sí               No         Fibra óptica   
5                   Sí               Sí         Fibra óptica   
8                   Sí               Sí         Fibra óptica   
13                  Sí               Sí         Fibra óptica   
26                  Sí               Sí         Fibra óptica   

   Protección contra Malware  ... Soporte técnico Televisión en streaming  \
4                         No  ...              No                      No   
5                         No  ...              No                      Sí   
8                         No  ...              Sí                      Sí   
13                        No  ...              No                      Sí   
26                        No  ...              No                      Sí   

   Películas en streaming   Contrato Facturación sin papel  \
4                      No  Mes a mes                    Sí   
5                      Sí  Mes a mes                    Sí   
8                      Sí  Mes a mes                    Sí   
13                     Sí  Mes a mes                    Sí   
26                     Sí  Mes a mes                    Sí   

                         Método de pago Cargos mensuales (dólares)  \
4                    Cheque electrónico                      70.70   
5                    Cheque electrónico                      99.65   
8                    Cheque electrónico                     104.80   
13  Transferencia bancaria (automática)                     103.70   
26                   Cheque electrónico                      99.35   

    Cargos totales  (dólares)  Deserción Cliente Cargos acumulados (dólares)  
4                      151.65                 Sí                      141.40  
5                      820.50                 Sí                      797.20  
8                     3046.05                 Sí                     2934.40  
13                    5036.30                 Sí                     5081.30  
26                    4749.15                 Sí                     4669.45  

[5 rows x 21 columns]

De los clientes con fibra óptica, 760 abandonaron la empresa.

6. 🤖 Utiliza la IA de Google Colab para realizar el código que permita responder esta pregunta:

Calcula el promedio de "Cargos mensuales" de los clientes que cumplen estas tres condiciones al mismo tiempo:

* Tienen servicio de internet por Fibra óptica.
* Usan protección contra Malware.
* Continúan en la empresa.

Luego, realiza el mismo cálculo para los clientes que cumplen estas otras tres condiciones:
* Tienen servicio de internet por DSL.
* No usan protección contra Malware.
* Han abandonado la empresa.

Finalmente, compara ambos promedios y responde: ¿Cuál grupo paga más en promedio por los cargos mensuales? ¿ A qué crees que se podría deber esa diferencia?

In [20]:
# prompt: Calcula el promedio de Cargos mensuales (dólares) de clientes cumplen estas tres condiciones a la vez: tienen servicio de internet por Fibra óptica, usan protección contra Malware y continúan en la empresa (Deserción cliente = No)

# Filter the DataFrame for customers with Fiber optic internet, Malware protection, and no churn
df_filtered_fiber_malware_no_churn = df_1[
    (df_1['Servicio de Internet'] == 'Fibra óptica') &
    (df_1['Protección contra Malware'] == 'Sí') &
    (df_1['Deserción Cliente'] == 'No')
]

# Calculate the average of 'Cargos mensuales (dólares)' for the first group
average_charges_fiber_malware_no_churn = df_filtered_fiber_malware_no_churn['Cargos mensuales (dólares)'].mean()

print(f"Promedio: ${average_charges_fiber_malware_no_churn:.2f}")

Promedio: $100.40


In [21]:
# prompt: Calcula el promedio de Cargos mensuales (dólares) de clientes cumplen estas tres condiciones a la vez: tienen servicio de internet por DSL, no usan protección contra Malware y no continúan en la empresa (Deserción cliente = Sí)

# Filter the DataFrame for customers with DSL internet, no Malware protection, and churn
df_filtered_dsl_no_malware_churn = df_1[
    (df_1['Servicio de Internet'] == 'DSL') &
    (df_1['Protección contra Malware'] == 'No') &
    (df_1['Deserción Cliente'] == 'Sí')
]

# Calculate the average of 'Cargos mensuales (dólares)' for the second group
average_charges_dsl_no_malware_churn = df_filtered_dsl_no_malware_churn['Cargos mensuales (dólares)'].mean()

print(f"Promedio: ${average_charges_dsl_no_malware_churn:.2f}")

Promedio: $45.86


El promedio de los cargos mensuales de los clientes que tienen servicio de internet por fibra óptica, cuentan con protección contra Malware y continúan en la compañía es de 100,4 dólares. Mientras que el promedio para los clientes que tienen servicio de internet por DSL, no cuentan con protección contra Malware y abandonaron la compañía es de 45,86 dólares.

El grupo que tiene un menor promedio de cargos mensuales es el segundo. Esta diferencia podría deberse a que los clientes del primer grupo están suscritos a planes más avanzados y costosos, como la fibra óptica y la protección contra Malware. Mientras que los clientes del segundo grupo estaban suscritos a planes más básicos y económicos, como el internet por DSL y la ausencia de protecciones adicionales, lo que podría reflejar una menor satisfacción con el servicio recibido.

